Prepare for java 8,11 3.0.0

leerho opened the pull request:
Prepares the java 8,11 code-base for a 3.0.0 release.

leerho committed:
Prepare for java 8,11 3.0.0

leerho committed:
Preparing for java 8/11 3.0.0

leerho committed:
aligning java 8/11 3.0.0 API & behavior to be as similar to what 4.0.0
will look like with java 17 FFM Panama

leerho committed:
Update.  Removing Handle references.

leerho committed:
Fixing a few minor things.

leerho committed:
Set poms to 3.0.0-SNAPSHOT

leerho committed:
Made corrections in javadocs, primarily.

Improved the toString methods.

Simplified BaseBuffer API.

leerho committed:
Changed name of BaseBuffer to Positional.

And BaseBufferImpl to PositionalImpl.

Cleaned up some duplicate methods.

Improved some javadocs.

leerho committed:
Renamed BaseBufferTest to PositionalTest

Corrected some tests that were referring to the wrong exception.

jmalkin commented on datasketches-memory-java8/src/main/java/org/apache/datasketches/memory/Buffer.java:
```diff
@@ -19,52 +19,52 @@
+   * The returned Buffer will use the <i>ByteOrder</i> of the <i>ByteBuffer</i>.
    * @param byteBuffer the given ByteBuffer, must not be null.
-   * @return a new Buffer for read-only operations on the given ByteBuffer.
+   * @return a new <i>Buffer</i> for read-only operations on the given ByteBuffer.
```
I believe `<code>` tags should be used instead of italics when referring to types

leerho replied:
You are correct and we have discussed this before.  However, this issue exists across the entire library, and we decided that if we were to change this style issue, we would need to do it for the entire library.  And I don't want to tackle it now.

jmalkin replied:
Yeah, my overall review comment was that we should not do that for this PR

jmalkin commented on datasketches-memory-java8/src/main/java/org/apache/datasketches/memory/MemoryRequestServer.java:
```diff
@@ -28,22 +28,31 @@
+   * determine the byte order of the returned WritableMemory and other properties.
    * @param currentWritableMemory the current writableMemory of the client. It must be non-null.
-   * @param capacityBytes The capacity being requested. It must be &ge; 0.
+   * @param newCapacityBytes The capacity being requested. It must be &gt; the capacity of the currentWritableMemory.
```
Why does it need to be larger?

leerho replied:
This is the MemoryRequestServer and the intent is to request a larger memory.  It doesn't make sense to request a memory space smaller than what you already have.

jmalkin replied:
So we don't support the pattern of getting a smaller amount of additional memory while keeping the current memory?

leerho replied:
Not with this interface, and never have.  If you want a separate smaller memory, use the regular allocate methods for that.  This is explicitly for the case where a sketch runs out of its allocated memory

jmalkin commented on datasketches-memory-java8/src/main/java/org/apache/datasketches/memory/Memory.java:
```diff
@@ -35,67 +33,76 @@
 /**
  * Defines the read-only API for offset access to a resource.
  *
- * <p>The classes in this package are not thread-safe.</p>
```
Are they now thread-safe?

leerho replied:
Rather than having this in every class, this comment applies to the whole package and this comment is already in the package-info file, which makes this statement redundant.

jmalkin commented on datasketches-memory-java8/src/main/java/org/apache/datasketches/memory/Memory.java:
```diff
@@ -35,67 +33,76 @@
-   * {@link #map(File, long, long, ByteOrder) map(file, 0, file.length(), ByteOrder.nativeOrder())}.
-   * @param file the given file to map. It must be non-null, length &ge; 0, and readable.
+   * {@link #map(File, long, long, ByteOrder)
+   * map(file, 0, file.length(), scope, ByteOrder.nativeOrder())}.
```
link has 4 parameters, example has 5. Make sure it points to the right place.

leerho replied:
Yep. this is a javadoc comment bug and it occurred in 3 places.  Will fix. I also searched the entire ds-memory repo for this bug and it is clean.

jmalkin commented on datasketches-memory-java8/src/main/java/org/apache/datasketches/memory/Memory.java:
```diff
@@ -35,67 +33,76 @@
-   * @param capacityBytes the size of the mapped memory. It must not be negative.
-   * @param byteOrder the byte order to be used for the mapped memory. It must be non-null.
+   * @param file the given file to map. It must be non-null, readable and length &ge; 0.
+   * @param fileOffsetBytes the position in the given file in bytes. It must &ge; 0.
```
must be &ge; 0

currently missing the word "be"

leerho replied:
will Fix.

jmalkin commented on datasketches-memory-java8/src/main/java/org/apache/datasketches/memory/WritableMemory.java:
```diff
@@ -33,120 +34,118 @@
-   * @param capacityBytes the size of the desired memory in bytes. It must be &ge; 0.
-   * @return WritableMemory for this off-heap resource.
+   * @param capacityBytes the size of the desired memory in bytes.
+   * @return WritableMemory for this off-heap, native resource.
```
I think it's worth specifying "byte order" explicitly so it's clear what "native" refers to.

leerho replied:
In this specific context I meant native=off-heap, but it is equally valid to mean native byte-order! This double use of "native" bugs me.  But the double use also exists in Panama.  I'm not sure what to do about this.  I could go through the entire repo and check for all uses of "native" and make sure the context is clear.

jmalkin commented on datasketches-memory-java8/src/main/java/org/apache/datasketches/memory/internal/ResourceImpl.java:
```diff
@@ -423,45 +430,58 @@ static final String toHex(final ResourceImpl state, final String preamble, final
-    sb.append("MemReq, hashCode    : ").append(memReqStr).append(LS);
-    sb.append("Valid               : ").append(state.isValid()).append(LS);
+    sb.append("MemReqSvr, hashCode : ").append(memReqStr).append(LS);
+    sb.append("Valid               : ").append(state.isAlive()).append(LS);
```
Should the text be changed to `Alive` to reflect the method name?

leerho replied:
Yes, and I will check the whole repo for this substitution.

jmalkin commented on datasketches-memory-java8/src/test/java/org/apache/datasketches/memory/internal/MemoryTest.java:
```diff
@@ -96,15 +96,15 @@ public void checkArrayWrap() {
     }
     // check 0 length array wraps
-    Memory memZeroLengthArrayBoolean = WritableMemory.writableWrap(new boolean[0]);
+    //Memory memZeroLengthArrayBoolean = WritableMemory.writableWrap(new boolean[0]);
```
should just remove these entirely instead of commenting them out

leerho replied:
Will remove.

jmalkin commented on datasketches-memory-java8/src/test/java/org/apache/datasketches/memory/internal/ExampleMemoryRequestServerTest.java:
```diff
@@ -0,0 +1,113 @@
+/**
+ * Example of how to use the MemoryRequestServer with a memory hungry client.
+ *
+ * <p>Note: this example only works with Java 17.</p>
```
Is this accurate? If so, is it excluded from running?

leerho replied:
Not accurate.  Removed.  It works fine in Java8

jmalkin reviewed (changes requested):
There are a few very, very minor things.

We should try to move all the `<i>` stuff to `<code>` but that's not a priority for this PR.

leerho committed:
For all public classes in /memory/ (and not /memory/internal/):

1. I have corrected the javadocs and code comments to use "off-heap"
instead of "direct" or "native". However, I decided not to change the
use of "Direct" in existing method calls or in the /internal/ directory.
That would cause a huge number of changes (100's).

2. I have corrected the javadocs and code comments to use "alive" or
"isAlive" instead of "valid" or "isValid". However, I decided not to
change the word "Valid" in the /internal/ directory.  That would cause a
huge number of changes (100's).

jmalkin reviewed (approved):
Thankfully a single commit so I was able to look at that diff rather than try to figure out what changed in the entire PR :)
